### ISSUE TITLE
work on new cleaning method

### ISSUE THREAD

**Issue by lizzieinvancouver** [alinazeng/Tree_Spotters#3]
Based on results from issue #2 I think we should adjust our cleaning methods in the following ways:

1. I think now 'clean' means the same observation by two different observers on the same day, right? If so, we should change this to on the same day or the day before or after (at least).
2. Take the 'clean' data and calculate the mean and total range, the allow in all 'unclean' observations that are within 2X of the range. For example, if the mean of the clean is 30, and the range is 23-40 (that's 17 days range), then we would take 30-17 and 30+17 (13 to 47) as the allowable range to include 'unclean' observations.
3. We could also set some a priori range -- such as, we take all observations within 2 weeks before and 2 weeks after the mean of the clean data ... 

Let me know if you want me to work on the code for this! If so, please point me to the file and file lines where the cleaning happens (Or you can take a stab at it and I can help of course.)

**Comment by alinazeng** [comment]
Hi Lizzie
 Thanks for the response. I think I can figure out the code :D If my understanding is correct, essentially we use the clean data and use it as the reference in the middle and accept any unclean data that isn't too extreme.  I will work on this on Thursday and write you in case I run into issues. Thank you!

**Comment by alinazeng** [comment]
Hi Lizzie,

I know it is a holiday so please do not respond: I am just putting down what I found so far. Feel free to reply when you have time next week.

Apparently, the so-called "clean" data we have is not too clean after all... I calculated the means and the ranges and quite a lot of the ranges are greater than 100... Please see [here](https://github.com/alinazeng/Tree_Spotters/blob/main/output/problematic_clean_range_July1.csv)

it might be safer for us to go with option 3 (setting a prior range)... Or we could take the range of the means across 6 years and use that value as our guide: for example, the mean bb doy of American basswood  is 117, 104,  119, 139, 115, 111 across 2015-2020, and the range is 139-104 = 365, which we could use on the unclean data (tho this probably isn't analytically correct)

let me know what you think~ Thank you so much!

**Comment by lizzieinvancouver** [comment]
@alinazeng Nice work on this! There are some alarming values in there. And actually some are affecting the mean (e.g., eastern cottonwood leafout in 2020 seems quite late at 189) so I think we may need a better approach than even option 3. 

Can you add the number of observations of the event going into the mean estimate (and the min/max)? I ask because I wonder if we should have a minimum number of observations before we calculate the mean.

**Comment by alinazeng** [comment]
@lizzieinvancouver Hi Lizzie. Please see the column number_obs in[ here](https://github.com/alinazeng/Tree_Spotters/blob/main/output/problematic_clean_range_July1.csv)

```
 range(summ_clean_year$number_obs)
[1]  1 36
```


```
 table(summ_clean_year$number_obs)

  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  23 
 94  64  47  77 122  68  60  42  36  27  17  14   7   3   4   4   1   2   5   2   6   2 
 27  36 
  1   1 
```

```
 median(summ_clean_year$number_obs)
[1] 5
```

Good point! now that I see the count we should defo set a min number of observations. What do you think? Thanks!

**Comment by lizzieinvancouver** [comment]
@alinazeng Hmm, well, it doesn't look like the number of observations is very correlated with how crazy the range and/or mean is! Or at least some of the more surprising values have 10 observations....

First, can you double-check that our cleaning code requires the same observation on the same day? Or you can point me to the code and I can check. I am just surprised we're having two observers say there is new leafout of American Basswood on day 209 ... I wonder if they are recording a SECOND leaf flush (that is, new leaves produced after spring leafout; trees do this sometimes). If possible, could you look at the data for that observer and see if leafout was previously recorded? If it was, one fix we should do before others (see below) is to only take the first observation of an event per observer per tree ID.

Either way, I am not sure there's a perfect option. I thought of setting some prior values and/or limits (for example, all trees must leafout before June, so throw out all observations of leafout in June), but I feel uncomfortable doing that because outliers can be real -- especially as our tree spotters look closely and the trees can experience frost, pruning etc.. which may delay them.

I think your idea of the range of average yearly values sounds good. However, mean values tend to be much less variable than the individual values (this is basically because of the Central Limit Theorem) so maybe we should do the interannual range times 2 (2X) or 3 (3X) ... so in your American basswood example the mean is 141 and the range is 35 (17.5 on either side of 141) we'd allow values of 106-176 if we did 2X, or 88.5-193.5 (141-52.5 to 141+52.5) for 3X.

**Comment by alinazeng** [comment]
@lizzieinvancouver 
Hi Lizzie,
After some looking around I spotted the problem. We do need an alternate cleaning system (this comment is quite long so bear with me please)

Below is what Cat used to clean 

`d <- d[(d$Multiple_FirstY>=1 | d$Multiple_Observers>0),] `     This selects data where multiple people observed the same phenophase

> **Multiple_FirstY**: Indicates whether there are multiple series for the phenophase on the organism within the selected time period. A value of "0" indicates there is only one series, and a value of "1" indicates there are more than one series within the selected time period. When the selected time period consists of more than one year, calculation of multiple series is separate for each 12-month time span starting from the selected start date. Note that the first "yes" record for a 12-month time span during the selected time period might not be preceded by a "no" record, but the first   "yes" for subsequent series within the same 12-month time span is always separated from the previous series by at least one "no" record.

> **Multiple_Observers**: Indicates whether multiple observers contributed records to the series. A value of "0" indicates only one observer contributed records, and a value of "1" indicates more than one observer contributed records.

`d <- d[(d$NumYs_in_Series>=3),] `    This selects data where the same phenophase was seen 3 times in a row
`d <- d[(d$NumDays_Since_Prior_No>=0 & d$NumDays_Since_Prior_No<=14),]   ` this limits to data where a no is followed by a yes, so that it is a new observation/new phenophase but has been detected within a reasonable timeframe

Re. American basswood bud burst on day 209 in 2019.
I went and checked, two American basswood trees (id %in% c(87757,86289)) were recorded to exhibit bb on day 209 by 3 Treespotters ('46197,19414,28846') see [here](https://github.com/alinazeng/Tree_Spotters/blob/main/output/observer_basswood2019_July2.csv) (one of them is our [keenest spotter](https://github.com/alinazeng/Tree_Spotters/blob/main/output/number_of_observation_by_each_individual_each_year_June_2.csv)). 

Judging from their difference in contribution, I did not think they were observing at the same time. I then checked the very raw data and found the issue (the three observations were indeed made at different times and only one person observed bb, the other 2 people noted bb as "0" (ie. not happening). 

![image](https://user-images.githubusercontent.com/83845945/124323776-b49f7c80-db36-11eb-8cbc-d5b2467b9938.png)

This is problematic (also good that we caught the problem) because apparently, the **Multiple_Observers** field we have been using and thinking it helped us get rid of observations that were made by only one folk is actually failing us :))) in the case of the American basswood, for example, because more than one person went to see the tree in question and submitted some observation about that tree, the  **Multiple_Observers** field auto-generated by NPN gives us the number 1 (meaning observed by multiple people) as the tree in question was indeed visited by more than one person, even though some people thought budburst is happening while others thought otherwise.

tldr: I need to clean the data from scratch as said issues probably are affecting other trees and phenophases too. I am happy to try things out and see how it goes. Let me know if you have any suggestions.
Thanks for tolerating this long-winded response!

**Comment by lizzieinvancouver** [comment]
@alinazeng Wow! That's not at all long winded for all the info you packed in there. Great sleuthing! 

It's somewhat bad news in that we need to build up some new cleaning code, but this is how good science goes. And it's super good news in that maybe we can come up with a better way to clean! Also, this will affect @cchambe12 current paper that she was planning to submit soon, so a super important finding -- thank you!

A few thoughts, I like these two lines of code:

> `d <- d[(d$NumYs_in_Series>=3),] ` This selects data where the same phenophase was seen 3 times in a row
> `d <- d[(d$NumDays_Since_Prior_No>=0 & d$NumDays_Since_Prior_No<=14),] ` this limits to data where a no is followed by a yes, so that it is a new observation/new phenophase but has been detected within a reasonable timeframe

And it seems like  `Multiple_FirstY` is doing some helpful work for us, by going year (12 month by 12 month) year.

I think the main thing we want to do now is: **Select events that are observed by at least 2 different observers** over, let's say, a FIVE DAY (5 day) timespan. I said 3 days before, but I think 5 days works. 

There's multiple ways to code this but I might try a loop that finds the first '1' per event per species per year and then searches that date, plus two days before and two days after **for different observers** and counts the other '1' ... there are probably more elegant ways to do this, but that is my first idea!

I think we can do it **per species** instead of per tree ID, as that will be more permissive. But let me know if you see an issue with that. 

We could also allow in all observations by certain observers, since we know them. For example, Beth and Kevin M. are very good, as is Suzanne M. ... that's not critical to do now I think, but something we could consider. We'd have to match up their names to their observer IDs. Step 1 on that would be to email the US NPN (cc me) and ask if the folks who observe can see their own ID, then we could ask Beth and Kevin, and Suz.

**Comment by alinazeng** [comment]
@lizzieinvancouver Hi Lizzie,

Thanks for being so open to finding a different cleaning approach! 

I thought of an alternate way to select data where multiple people observed the same phenophase.
Since we have the status number 1s and 0s representing the occurrence or absence of an event, we can `group_by` observations by `year, phase, tree ID, doy` and sum up the status -> a sum of 2 or above will, in theory, confirm the occurrence of that phenophase on that day (unless two people made the same wrong judgement in which case we can increase the limit to 3).

Let me know how this sounds. I will try this out tomorrow and look for ways to accept events observed by at least 2 different observers over a 5-day timespan. 

PS. I think we need to keep it as *per tree ID* until after we get rid of unreliable observations, otherwise the same incorrect observations made by one Tree Spotter will be added up which will lead to incorrect selection (hope this makes sense). We can go ahead with per species after some basic cleaning.

**Comment by lizzieinvancouver** [comment]
@alinazeng That's a great approach! I think the only trick will be how to extend it to more days, as once an event switches to 1, it stays 1. However, your approach could work on more days if you first create a dataset that extracts only the first 1 for each `year, phase, tree ID` -- then you can sum within some defined window and your idea will still work. 

Good point about the per tree ID -- thanks for thinking this all through so clearly!

**Comment by alinazeng** [comment]
Hi Lizzie @lizzieinvancouver 

Please see here for some subsets of raw data I organized
- [ doy of all "Yes" observations](https://github.com/alinazeng/Tree_Spotters/blob/main/output/observations_including_single_observers_July6.csv)
- [doy of "Yes" observations shared by at least two people](https://github.com/alinazeng/Tree_Spotters/blob/main/output/observations_excluding_single_observers_July6.csv)

(obs_number: number of observations)

My thoughts are
- we probably should leave out the 2015 observations cuz they are pretty limited
- there are a few ways we could approach this:
1. we can use the mean of doy per species per year calculated from observations made by more than one person and use that as the guideline to get rid of crazy observations made by some individuals....(excuse the phrasing) the interquartile range, in some cases, might be of help too
2. make sure the first_doy and last_doy are either
- observed by more than one person (obs_number >1), or
- are seconded by a different observer within a 5 day period -> this is where I can't wrap my head around: how could I tell R to look at Row A's relation to other rows (essentially telling R to categorize observations as valid if there are known records of observations of the same phase on the same tree individual with a doy that is within doy of Row A plus or minus 5.

Thanks for your ongoing support!

**Comment by lizzieinvancouver** [comment]
Yes, agreed on 2015! It was a 'test' year so okay to leave out.

There may be a better way to do this but I would set up a loop and use the` %in% `command to sum observations in some date range ...

Assume you have a dataframe (`df`) each stage by year by tree ID (obsYESNO is the 0/1 column):

```
dayrange <- 5
goodrows <- c() # empty vector to use in loop
for (rownum in c(1:nrow(df)){
    dayz <- seq(df$doy[i]-dayrange, df$doy[i]+dayrange) # set up a vector of allowed days
    sumhere <- sum(df$obsYESNO[which(df$doy %in% dayz)]) # this should sum up observations in that range, indexing one column based on conditions in another column
    if(sumhere>1){ 
        goodrows <- c(goodrows, i) # trying to record which rows > 1 [see](https://www.datamentor.io/r-programming/if-else-statement/) though there is probably a nicer way to add this; you could also write out a dataframe of the sums so you can see how many are 1, 2, 3 or more.
    }
```

Also, it seems like you already did this, but you can get all first YES events by taking the minimum... something like

```
## summarizing data 
allfirsts <-
      ddply(dataframe, c("year", "phenophase", "treeID", "observer"), summarise,
      firstevent = min(doy))
```

**Comment by alinazeng** [comment]
Hi Lizzie,

Thanks for the code. I tried it several times but some steps seem to be missing. 
My questions is: are we applying the loop on each row of a single dataframe, or are we splitting a dataframe into lists first before running the loop? I thought `[i]` is used on different items in the same list, but we have not split the dataframe into lists just yet. Please let me know! Thank you!

**Comment by lizzieinvancouver** [comment]
@alinazeng Yes! It is skeleton code, sorry about that. 

The loop should be applied on each row of a single dataframe (so` i `goes through each row). I might create a new extraID column of treeID x phase x year and make an outside loop (outside the one I wrote above) that makes a dataframe by subsetting to each unique value in that extraID column. So the outer loop would build mini dataframe (subsetted to each treeID x phase x year) and the inner loop would go through each row. Let me know if you want me to look at anything. For loops are a semi-advanced topic! But super useful.

**Comment by alinazeng** [comment]
Thanks lizzie. I will keep trying. The issue right now is the data frame is too big and takes forever to run. I will experiment with a subset first. Talk to you soon.

**Comment by lizzieinvancouver** [comment]
@alinazeng Hmm, yes, the bad things about loops is that they are very slow in R, but that is okay if we only need to run the code every so often. Always good to do a small test case before building up!

**Comment by alinazeng** [comment]
@lizzieinvancouver Hi Lizzie,

I am putting the loop function on pause at the moment cuz I ran into a rabbit hole lol (the day range does not seem to work, perhaps I'm not doing it correctly).

However, I came up with a *creative* workaround which seemed to work very well (let me know if you see an issue with this)

In a gist, I ordered the doy per year per tree_id per phase, and I take the difference of the doys

```
df <- df   %>% 
  group_by(year, phase, tree_id) %>%  
  arrange(doy, .by_group = TRUE)   # ascend doy based on groups 
```
```

df <- df  %>% group_by(year, phase, tree_id) %>%
  mutate(difference = doy - lag(doy,default=first(doy)))  # calculate the difference between doys
```

And of course, I then descended doys based on groups and calculated the difference the other way around, and created two fields : `difference` and `difference_reverse`  see[ here](https://github.com/alinazeng/Tree_Spotters/blob/main/output/doygaps.csv)

Since we were trying to accept single observations that are within a 5 day timeframe, the difference we just calculated can help us exclude ones that are observed at weird times

```
cleandf<- filter(df,difference < 5 | obs_num > 1) # this selects observations with a difference within 5 OR observations made by at least 2 people`
cleandf <- cleandf[!(cleandf$difference == 0 & cleandf$difference_reverse >5 & cleandf$obs_num==1),]    # this gets rid of first doys observed by a single observer that are too far from the rest of observations 
```

Please see [here](https://github.com/alinazeng/Tree_Spotters/blob/main/output/cleaningJuly07.csv) for the dataframe we have after my proposed cleaning. The numbers seem pretty reasonable. Let me know what you think.

 
PS. something else I realized after seeing the raw data is that for most of the less popular phases (seed drop, ripe fruit, pollen) we only have 1 observation per tree per year (if not none), so there's no way to do an average or extrapolate a time span.

**Comment by lizzieinvancouver** [comment]
@alinazeng That's a great alternative approach! Better than mine I think. I like it! And I think the numbers look good. 

Now, what to do about phases with only one observation per tree ID per year ... we could make one veryclean dataset with the method we've worked and make a lessclean dataset where we do the same but by tree species? Any ideas?

Thanks again for your great work on this!

**Comment by alinazeng** [comment]
@lizzieinvancouver hahahha thanks Lizzie.

Regarding those trees with only one observation, i forgot to mention to you, those single observations are oftentimes made by the same Tree Spotter on a single day, meaning that all the tree individuals of the same species have the same doy record (as they were visited by the same person at the same time)

however there are some species with abundant flower and fruit observations -> for the document I will likely pick several species for which we have enough data, and talk about the lack of observation for other species in the discussion. Does this sound good?

**Comment by lizzieinvancouver** [comment]
@alinazeng Could you add a wrap-up comment here about how you ended up cleaning the data and link to the script? That way in the future, folks can quickly scroll to the bottom and see what we did. 

Thank you!

**Comment by alinazeng** [comment]
morning Lizzie. Sounds good. Will do in a bit!

**Comment by alinazeng** [comment]
Recap of cleaning method:

I ordered the doy per year per tree_id per phase, and I take the difference of the doys

```
df <- df   %>% 
  group_by(year, phase, tree_id) %>%  
  arrange(doy, .by_group = TRUE)   # ascend doy based on groups 
```
```

df <- df  %>% group_by(year, phase, tree_id) %>%
  mutate(difference = doy - lag(doy,default=first(doy)))  # calculate the difference between doys
```

And of course, I then descended doys based on groups and calculated the difference the other way around, and created two fields : `difference` and `difference_reverse`  see[ here](https://github.com/alinazeng/Tree_Spotters/blob/main/output/doygaps.csv)

Since we were trying to accept single observations that are within a 5 day timeframe, the difference we just calculated can help us exclude ones that are observed at weird times

```
cleandf<- filter(df,difference < 5 | obs_num > 1) # this selects observations with a difference within 5 OR observations made by at least 2 people`
cleandf <- cleandf[!(cleandf$difference == 0 & cleandf$difference_reverse >5 & cleandf$obs_num==1),]    # this gets rid of first doys observed by a single observer that are too far from the rest of observations 
```

link to the script
https://github.com/alinazeng/Tree_Spotters/blob/065c4c49dd9ad65586f7833657b71ea9a375472b/analyses/process_work_alina/script_to_clean_differently_July1_obselete.R#L214-L258

Please reach Alina at alinazengziyun@yahoo.com if you have any questions

**Comment by lizzieinvancouver** [comment]
@alinazeng Wonderful! Thank you!
